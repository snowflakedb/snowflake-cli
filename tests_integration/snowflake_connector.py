# Copyright (c) 2024 Snowflake Inc.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
# http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

from __future__ import annotations

import os
import uuid
from contextlib import contextmanager
from typing import Optional
from unittest import mock

import pytest
from snowflake import connector
from snowflake.cli.api.exceptions import EnvironmentVariableNotFoundError
from snowflake.cli._app.snow_connector import update_connection_details_with_private_key

_ENV_PARAMETER_PREFIX = "SNOWFLAKE_CONNECTIONS_INTEGRATION"
SCHEMA_ENV_PARAMETER = f"{_ENV_PARAMETER_PREFIX}_SCHEMA"
DATABASE_ENV_PARAMETER = f"{_ENV_PARAMETER_PREFIX}_DATABASE"


def add_uuid_to_name(name: str) -> str:
    return f"{name}_{uuid.uuid4().hex}"


@contextmanager
def mock_single_env_var(name: str, value: str):
    env = dict(os.environ)
    env[name] = value
    with mock.patch.dict(os.environ, env):
        yield


def _escape_name(name: str) -> str:
    if "-" in name:
        name = f'"{name}"'
    return name


@contextmanager
def setup_test_database(snowflake_session, database_name: str):
    database_name = _escape_name(database_name)
    snowflake_session.execute_string(
        f"create database {database_name}; use database {database_name}; use schema public;"
    )
    with mock_single_env_var(DATABASE_ENV_PARAMETER, value=database_name):
        yield
    snowflake_session.execute_string(f"drop database {database_name}")


@contextmanager
def setup_test_schema(snowflake_session, schema_name: str):
    schema_name = _escape_name(schema_name)
    snowflake_session.execute_string(
        f"create schema {schema_name}; use schema {schema_name};"
    )
    with mock_single_env_var(SCHEMA_ENV_PARAMETER, value=schema_name):
        yield
    snowflake_session.execute_string(f"drop schema {schema_name}")


@pytest.fixture(scope="function")
def test_database(snowflake_session):
    database_name = add_uuid_to_name("db")
    with setup_test_database(snowflake_session, database_name):
        yield database_name


@pytest.fixture(scope="function")
def test_role(snowflake_session):
    role_name = f"role_{uuid.uuid4().hex}"
    snowflake_session.execute_string(
        f"create role {role_name}; grant role {role_name} to user {snowflake_session.user};"
    )
    yield role_name
    snowflake_session.execute_string(f"drop role {role_name}")


@pytest.fixture(scope="session")
def snowflake_session():
    config = {
        "application": "INTEGRATION_TEST",
        "authenticator": "SNOWFLAKE_JWT",
        "account": _get_from_env("ACCOUNT"),
        "user": _get_from_env("USER"),
        "private_key_file": _get_private_key_file(),
        "private_key_raw": _get_from_env("PRIVATE_KEY_RAW", allow_none=True),
        "host": _get_from_env("HOST", allow_none=True),
        "warehouse": _get_from_env("WAREHOUSE", allow_none=True),
        "role": _get_from_env("ROLE", allow_none=True),
    }
    config = {k: v for k, v in config.items() if v is not None}
    update_connection_details_with_private_key(config)
    connection = connector.connect(**config)
    yield connection
    connection.close()


def _get_from_env(parameter_name: str, default=None, allow_none=False) -> str | None:
    env_value = os.environ.get(f"{_ENV_PARAMETER_PREFIX}_{parameter_name}")
    if not env_value:
        if default is None and not allow_none:
            raise EnvironmentVariableNotFoundError(
                f"{_ENV_PARAMETER_PREFIX}_{parameter_name}"
            ) from None
        return default
    return env_value


def _get_private_key_file() -> Optional[str]:
    private_key_file = _get_from_env("PRIVATE_KEY_PATH", allow_none=True)
    if private_key_file is not None:
        return private_key_file
    return _get_from_env("PRIVATE_KEY_FILE", allow_none=True)
