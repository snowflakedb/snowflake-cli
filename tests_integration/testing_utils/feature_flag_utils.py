import os
from contextlib import contextmanager
from unittest import mock

from snowflake.cli.api.feature_flags import FeatureFlagMixin


@contextmanager
def with_feature_flags(flags: dict[FeatureFlagMixin, bool]):
    with mock.patch.dict(
        "os.environ",
        os.environ.copy()
        | {flag.env_variable(): str(value) for flag, value in flags.items()},
    ):
        yield
