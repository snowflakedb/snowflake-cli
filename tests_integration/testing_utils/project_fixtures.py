from textwrap import dedent

import pytest

from tests.nativeapp.factories import (
    ProjectV2Factory,
    ApplicationPackageEntityModelFactory,
    ApplicationEntityModelFactory,
    ManifestFactory,
)


MANIFEST_BASIC = ManifestFactory()

PYTHON_W_SNOWPARK = dedent(
    """\
    from snowflake.snowpark.functions import udf
    @udf(
        name="echo_fn",
        native_app_params={"schema": "core", "application_roles": ["app_public"]},
    )
    def echo_fn(data: str) -> str:
        return "echo_fn: " + data
        """
)


@pytest.fixture
def setup_v2_project_w_subdir(temp_dir):
    def wrapper():
        readme_v1 = (
            "This is the <% ctx.pkg_v1.stage_subdirectory %> version of this package!"
        )
        readme_v2 = (
            "This is the <% ctx.pkg_v2.stage_subdirectory %> version of this package!"
        )
        project_name = "stage_w_subdirs"
        ProjectV2Factory(
            pdf__entities=dict(
                pkg_v1=ApplicationPackageEntityModelFactory(
                    identifier=f"<% fn.concat_ids('{project_name}_pkg_', ctx.env.USER) %>",
                    manifest="",
                    artifacts=[{"src": "app/v1/*", "dest": "./"}],
                    stage_subdirectory="v1",
                ),
                app_v1=ApplicationEntityModelFactory(
                    fromm__target="pkg_v1",
                    identifier=f"<% fn.concat_ids('{project_name}_app_v1_', ctx.env.USER) %>",
                ),
                pkg_v2=ApplicationPackageEntityModelFactory(
                    identifier=f"<% fn.concat_ids('{project_name}_pkg_', ctx.env.USER) %>",
                    manifest="",
                    artifacts=[{"src": "app/v2/*", "dest": "./"}],
                    stage_subdirectory="v2",
                ),
                app_v2=ApplicationEntityModelFactory(
                    fromm__target="pkg_v2",
                    identifier=f"<% fn.concat_ids('{project_name}_app_v2_', ctx.env.USER) %>",
                ),
            ),
            files={
                "app/v1/manifest.yml": MANIFEST_BASIC,
                "app/v1/README.md": readme_v1,
                "app/v1/setup.sql": "SELECT 1;",
                "app/v2/manifest.yml": MANIFEST_BASIC,
                "app/v2/README.md": readme_v2,
                "app/v2/setup.sql": "SELECT 1;",
            },
        )
        return project_name, temp_dir

    return wrapper


@pytest.fixture
def setup_v2_project_w_subdir_w_snowpark(temp_dir):
    def wrapper():
        setup_script = dedent(
            """\
                CREATE APPLICATION ROLE IF NOT EXISTS app_public;
                CREATE OR ALTER VERSIONED SCHEMA core;
                GRANT USAGE ON SCHEMA core TO APPLICATION ROLE app_public;
                """
        )
        project_name = "stage_w_subdirs"
        ProjectV2Factory(
            pdf__entities=dict(
                pkg_v1=ApplicationPackageEntityModelFactory(
                    identifier=f"<% fn.concat_ids('{project_name}_pkg_', ctx.env.USER) %>",
                    manifest="",
                    artifacts=[
                        {"src": "app/v1/*", "dest": "./", "processors": ["snowpark"]}
                    ],
                    stage_subdirectory="v1",
                ),
                app_v1=ApplicationEntityModelFactory(
                    fromm__target="pkg_v1",
                    identifier=f"<% fn.concat_ids('{project_name}_app_v1_', ctx.env.USER) %>",
                ),
                pkg_v2=ApplicationPackageEntityModelFactory(
                    identifier=f"<% fn.concat_ids('{project_name}_pkg_', ctx.env.USER) %>",
                    manifest="",
                    artifacts=[
                        {"src": "app/v2/*", "dest": "./", "processors": ["snowpark"]}
                    ],
                    stage_subdirectory="v2",
                ),
                app_v2=ApplicationEntityModelFactory(
                    fromm__target="pkg_v2",
                    identifier=f"<% fn.concat_ids('{project_name}_app_v2_', ctx.env.USER) %>",
                ),
            ),
            files={
                "app/v1/manifest.yml": MANIFEST_BASIC,
                "app/v1/README.md": "\n",
                "app/v1/setup.sql": setup_script,
                "app/v1/module-echo-v1/echo-v1.py": PYTHON_W_SNOWPARK,
                "app/v2/manifest.yml": MANIFEST_BASIC,
                "app/v2/README.md": "\n",
                "app/v2/setup.sql": setup_script,
                "app/v2/module-echo-v2/echo-v2.py": PYTHON_W_SNOWPARK,
            },
        )
        return project_name, temp_dir

    return wrapper
