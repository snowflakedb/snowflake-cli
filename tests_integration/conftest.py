# Copyright (c) 2024 Snowflake Inc.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
# http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

from __future__ import annotations

import functools
import json
import os
import shutil
import tempfile
from contextlib import contextmanager
from dataclasses import dataclass
from json import JSONDecodeError
from pathlib import Path
from typing import Any, Dict, List, Optional
from unittest import mock
from uuid import uuid4

import pytest
import yaml
from typer import Typer
from typer.testing import CliRunner

from snowflake.cli._app.cli_app import CliAppFactory
from snowflake.cli.api.cli_global_context import (
    fork_cli_context,
    get_cli_context_manager,
)
from snowflake.cli.api.connections import OpenConnectionCache
from snowflake.cli.api.project.util import TEST_RESOURCE_SUFFIX_VAR
from tests.conftest import clean_logging_handlers_fixture  # noqa: F401
from tests.testing_utils.files_and_dirs import merge_left
from tests.testing_utils.fixtures import (
    alter_snowflake_yml,  # noqa: F401
    snowflake_home,
)

pytest_plugins = [
    "tests_common",
    "tests_common.deflake",
    "tests_integration.testing_utils",
    "tests_integration.snowflake_connector",
]

TEST_DIR = Path(__file__).parent
DEFAULT_TEST_CONFIG = "connection_configs.toml"
WORLD_READABLE_CONFIG = "world_readable.toml"
IS_QA = "qa" in os.getenv("SNOWFLAKE_CONNECTIONS_INTEGRATION_HOST", "").lower()


@dataclass
class CommandResult:
    exit_code: int
    json: Optional[List[Dict[str, Any]] | Dict[str, Any]] = None
    output: Optional[str] = None
    stderr: Optional[str] = None


class TestConfigProvider:
    def __init__(self, temp_dir_with_configs: Path):
        self._temp_dir_with_configs = temp_dir_with_configs

    def get_config_path(self, file_name: str) -> Path:
        return self._temp_dir_with_configs / file_name


@pytest.fixture(scope="session")
def test_snowcli_config_provider():
    with tempfile.TemporaryDirectory() as td:
        temp_dst = Path(td) / "config"
        shutil.copytree(TEST_DIR / "config", temp_dst)
        for config_file in temp_dst.glob("**/*.toml"):
            if config_file.name != WORLD_READABLE_CONFIG:
                config_file.chmod(0o600)  # Make config file private
        yield TestConfigProvider(temp_dst)


@pytest.fixture(scope="session")
def test_root_path():
    return TEST_DIR


class SnowCLIRunner(CliRunner):
    def __init__(
        self,
        app: Typer,
        test_config_provider: TestConfigProvider,
        default_username: str,
        resource_suffix: str,
    ):
        super().__init__()
        self.app = app
        self._test_config_provider = test_config_provider
        self._test_config_path = self._test_config_provider.get_config_path(
            DEFAULT_TEST_CONFIG
        )
        self._default_username = default_username
        self._resource_suffix = resource_suffix

    def use_config(self, config_file_name: str) -> None:
        self._test_config_path = self._test_config_provider.get_config_path(
            config_file_name
        )

    @functools.wraps(CliRunner.invoke)
    def invoke(self, *a, **kw):
        if "catch_exceptions" not in kw:
            kw.update(catch_exceptions=False)
        kw = self._with_env_vars(kw)

        # between every invocation, we need to reset the CLI context
        # and ensure no connections are cached going forward (to prevent
        # test cases from impacting each other / align with CLI usage)
        with fork_cli_context():
            connection_cache = OpenConnectionCache()
            cli_context_manager = get_cli_context_manager()
            cli_context_manager.reset()
            cli_context_manager.connection_cache = connection_cache
            try:
                return super().invoke(self.app, *a, **kw)
            finally:
                connection_cache.clear()

    def _with_env_vars(self, kw) -> dict:
        """
        Add required env vars to the invocation context if necessary and return new kwargs.

        Sets the USER env var to a default value if not set in the test,
        to allow us to use <% ctx.env.USER %> in test data on Windows.

        Sets the resource suffix env var unconditionally.
        The CLI automatically appends the value of this env var to some
        created resource identifiers, let's use this behaviour to add a unique
        suffix to resources used in tests to allow us to run simultaneous instances.
        """
        env = kw.get("env", {})
        return {
            **kw,
            "env": {
                **env,
                "USER": env.get("USER", self._default_username),
                TEST_RESOURCE_SUFFIX_VAR: self._resource_suffix,
            },
        }

    def invoke_with_config(self, args, **kwargs) -> CommandResult:
        result = self.invoke(
            [
                "--config-file",
                self._test_config_path,
                *args,
            ],
            **kwargs,
        )

        if result.output == "" or result.output.strip() == "Done":
            return CommandResult(result.exit_code, json=[])
        try:
            return CommandResult(
                result.exit_code, json.loads(result.output), output=result.output
            )
        except JSONDecodeError:
            return CommandResult(result.exit_code, output=result.output)

    def invoke_json(self, args, **kwargs) -> CommandResult:
        return self.invoke_with_config([*args, "--format", "JSON"], **kwargs)

    def invoke_with_connection_json(
        self, args, connection: str = "integration", **kwargs
    ) -> CommandResult:
        return self.invoke_json([*args, "-c", connection], **kwargs)

    def invoke_with_connection(
        self, args, connection: str = "integration", **kwargs
    ) -> CommandResult:
        return self.invoke_with_config([*args, "-c", connection], **kwargs)


@pytest.fixture
def runner(test_snowcli_config_provider, default_username, resource_suffix):
    app = CliAppFactory().create_or_get_app()
    yield SnowCLIRunner(
        app,
        test_snowcli_config_provider,
        default_username,
        resource_suffix,
    )


class QueryResultJsonEncoderError(RuntimeError):
    def __init__(self, output: str):
        super().__init__(f"Can not parse query result:\n{output}")


@pytest.fixture
def project_directory(temporary_working_directory, test_root_path):
    @contextmanager
    def _temporary_project_directory(
        project_name,
        merge_project_definition: Optional[dict] = None,
        subpath: Optional[Path] = None,
    ):
        test_data_file = test_root_path / "test_data" / "projects" / project_name
        project_dir = temporary_working_directory
        if subpath:
            project_dir = temporary_working_directory / subpath
            project_dir.mkdir(parents=True)
        shutil.copytree(test_data_file, project_dir, dirs_exist_ok=True)
        if merge_project_definition:
            with Path("snowflake.yml").open("r") as fh:
                project_definition = yaml.safe_load(fh)
            merge_left(project_definition, merge_project_definition)
            with open(Path(project_dir) / "snowflake.yml", "w") as file:
                yaml.dump(project_definition, file)

        yield project_dir

    return _temporary_project_directory


@pytest.fixture(autouse=True)
def reset_global_context_after_each_test(request):
    get_cli_context_manager().reset()
    yield


# This automatically used fixture isolates default location
# of config files from user's system.
@pytest.fixture(autouse=True)
def isolate_snowflake_home(snowflake_home):
    yield snowflake_home


@pytest.fixture
def default_username():
    return "snowflake"


@pytest.fixture
def resource_suffix(request):
    """
    Generate a random identifier suffix that includes the current test name.

    This suffix will be added to certain created resources like Native App
    packages and applications to be able to detect tests that don't
    clean up properly. The UUID is to avoid conflicts between concurrent runs.
    """
    # To generate a suffix that isn't too long or complex, we use originalname, which is the
    # "bare" test function name, without filename, class name, or parameterization variables
    return f"_{uuid4().hex}_{request.node.originalname}"


@pytest.fixture
def enable_snowpark_glob_support_feature_flag():
    with mock.patch(
        f"snowflake.cli.api.feature_flags.FeatureFlag.ENABLE_SNOWPARK_GLOB_SUPPORT.is_enabled",
        return_value=True,
    ), mock.patch(
        f"snowflake.cli.api.feature_flags.FeatureFlag.ENABLE_SNOWPARK_GLOB_SUPPORT.is_disabled",
        return_value=False,
    ):
        yield
