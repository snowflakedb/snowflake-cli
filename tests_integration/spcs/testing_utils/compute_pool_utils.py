# Copyright (c) 2024 Snowflake Inc.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
# http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

import math
import time

import pytest
from snowflake.connector import SnowflakeConnection

from tests_integration.conftest import SnowCLIRunner
from tests_integration.test_utils import contains_row_with, not_contains_row_with
from tests_integration.testing_utils.assertions.test_result_assertions import (
    assert_that_result_is_successful_and_executed_successfully,
    assert_that_result_is_successful_and_output_json_contains,
    assert_that_result_is_successful_and_output_json_equals,
)


class ComputePoolTestSetup:
    def __init__(
        self,
        runner: SnowCLIRunner,
        snowflake_session: SnowflakeConnection,
    ):
        self.runner = runner
        self.snowflake_session = snowflake_session


class ComputePoolTestSteps:
    def __init__(self, setup: ComputePoolTestSetup):
        self._setup = setup

    def create_compute_pool(self, compute_pool_name: str) -> None:
        result = self._setup.runner.invoke_with_connection_json(
            [
                "spcs",
                "compute-pool",
                "create",
                compute_pool_name,
                "--min-nodes",
                1,
                "--family",
                "CPU_X64_XS",
            ]
        )
        assert result.json, result.output
        assert "status" in result.json
        assert (
            f"Compute pool {compute_pool_name.upper()} successfully created."
            in result.json["status"]  # type: ignore
            or f"Compute pool {compute_pool_name.upper()} successfully created."
            in result.json["status"]  # type: ignore
        )

    def list_should_return_compute_pool(self, compute_pool_name) -> None:
        result = self._execute_list()
        assert_that_result_is_successful_and_output_json_contains(
            result, {"name": compute_pool_name.upper()}
        )

    def list_should_not_return_compute_pool(self, compute_pool_name: str) -> None:
        result = self._execute_list()
        assert not_contains_row_with(result.json, {"name": compute_pool_name.upper()})

    def describe_should_return_compute_pool(self, compute_pool_name: str) -> None:
        result = self._execute_describe(compute_pool_name)
        assert_that_result_is_successful_and_output_json_contains(
            result, {"name": compute_pool_name.upper()}
        )

    def status_should_return_compute_pool_idle_status(
        self, compute_pool_name: str
    ) -> None:
        result = self._setup.runner.invoke_with_connection_json(
            ["spcs", "compute-pool", "status", compute_pool_name]
        )
        assert_that_result_is_successful_and_output_json_contains(
            result, {"status": "IDLE"}
        )

    def drop_compute_pool(self, compute_pool_name: str) -> None:
        result = self._setup.runner.invoke_with_connection_json(
            [
                "object",
                "drop",
                "compute-pool",
                compute_pool_name,
            ],
        )
        assert_that_result_is_successful_and_output_json_equals(
            result, [{"status": f"{compute_pool_name.upper()} successfully dropped."}]
        )

    def stop_all_on_compute_pool(self, compute_pool_name: str) -> None:
        result = self._setup.runner.invoke_with_connection_json(
            ["spcs", "compute-pool", "stop-all", compute_pool_name]
        )
        assert_that_result_is_successful_and_executed_successfully(result, is_json=True)

    def suspend_compute_pool(self, compute_pool_name: str) -> None:
        result = self._setup.runner.invoke_with_connection_json(
            ["spcs", "compute-pool", "suspend", compute_pool_name]
        )
        assert_that_result_is_successful_and_executed_successfully(result, is_json=True)

    def resume_compute_pool(self, compute_pool_name: str) -> None:
        result = self._setup.runner.invoke_with_connection_json(
            ["spcs", "compute-pool", "resume", compute_pool_name]
        )
        assert_that_result_is_successful_and_executed_successfully(result, is_json=True)

    def set_unset_compute_pool_property(self, compute_pool_name: str) -> None:
        comment = "test comment"
        set_result = self._setup.runner.invoke_with_connection_json(
            ["spcs", "compute-pool", "set", compute_pool_name, "--comment", comment]
        )
        assert_that_result_is_successful_and_executed_successfully(
            set_result, is_json=True
        )

        describe_result = self._execute_describe(compute_pool_name)
        assert_that_result_is_successful_and_output_json_contains(
            describe_result, {"comment": comment}
        )
        unset_result = self._setup.runner.invoke_with_connection_json(
            ["spcs", "compute-pool", "unset", compute_pool_name, "--comment"]
        )
        assert_that_result_is_successful_and_executed_successfully(
            unset_result, is_json=True
        )
        describe_result = self._execute_describe(compute_pool_name)
        assert_that_result_is_successful_and_output_json_contains(
            describe_result, {"comment": None}
        )

    def wait_until_compute_pool_is_idle(self, compute_pool_name: str) -> None:
        self._wait_until_compute_pool_reaches_state(compute_pool_name, "IDLE", 900)

    def wait_until_compute_pool_is_suspended(self, compute_pool_name: str) -> None:
        self._wait_until_compute_pool_reaches_state(compute_pool_name, "SUSPENDED", 60)

    def _wait_until_compute_pool_reaches_state(
        self, compute_pool_name: str, target_state: str, max_duration: int
    ):
        assert max_duration > 0
        max_counter = math.ceil(max_duration / 10)
        target_state = target_state.upper()
        for i in range(max_counter):
            status = self._execute_describe(compute_pool_name)
            if contains_row_with(status.json, {"state": target_state}):
                return
            time.sleep(10)
        status = self._execute_describe(compute_pool_name)

        error_message = f"Compute pool {compute_pool_name} didn't reach target state '{target_state}' in {max_duration} seconds. Current state is '{status.json[0]['state']}'"
        pytest.fail(error_message)

    def _execute_describe(self, compute_pool_name: str):
        return self._setup.runner.invoke_with_connection_json(
            ["object", "describe", "compute-pool", compute_pool_name]
        )

    def _execute_list(self):
        return self._setup.runner.invoke_with_connection_json(
            [
                "object",
                "list",
                "compute-pool",
            ],
        )
