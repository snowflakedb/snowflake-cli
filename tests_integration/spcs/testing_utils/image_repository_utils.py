from snowflake.connector import SnowflakeConnection

from tests_integration.conftest import SnowCLIRunner
from tests_integration.testing_utils.assertions.test_result_assertions import (
    assert_that_result_is_successful_and_output_json_contains,
)


class ImageRepositoryTestSetup:
    def __init__(
        self,
        runner: SnowCLIRunner,
        snowflake_session: SnowflakeConnection,
    ):
        self.runner = runner
        self.snowflake_session = snowflake_session


class ImageRepositoryTestSteps:
    def __init__(self, setup: ImageRepositoryTestSetup):
        self._setup = setup

    def deploy_image_repository(self, image_repository_name: str) -> None:
        result = self._setup.runner.invoke_with_connection_json(
            [
                "spcs",
                "image-repository",
                "deploy",
            ]
        )
        assert_that_result_is_successful_and_output_json_contains(
            result,
            {
                "message": f"Image repository '{image_repository_name}' successfully deployed."
            },
        )

    def deploy_image_repository_with_replace(self, image_repository_name: str) -> None:
        result = self._setup.runner.invoke_with_connection_json(
            [
                "spcs",
                "image-repository",
                "deploy",
                "--replace",
            ]
        )
        assert_that_result_is_successful_and_output_json_contains(
            result,
            {
                "message": f"Image repository '{image_repository_name}' successfully deployed."
            },
        )
