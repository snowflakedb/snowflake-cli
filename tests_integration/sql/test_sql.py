# Copyright (c) 2024 Snowflake Inc.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
# http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

import time
from unittest import mock

import pytest

from tests_integration.testing_utils import ObjectNameProvider


@pytest.mark.integration
def test_query_parameter(runner):
    result = runner.invoke_with_connection_json(["sql", "-q", "select pi()"])

    assert result.exit_code == 0
    assert _round_values(result.json) == [{"PI()": 3.14}]


@pytest.mark.integration
def test_multi_queries_from_file(runner, test_root_path):
    result = runner.invoke_with_connection_json(
        [
            "sql",
            "-f",
            f"{test_root_path}/test_data/sql_multi_queries.sql",
        ]
    )

    assert result.exit_code == 0
    assert _round_values_for_multi_queries(result.json) == [
        [{"ROUND(LN(1), 4)": 0.0}],
        [{"ROUND(LN(10), 4)": 2.3}],
        [{"ROUND(LN(100), 4)": 4.61}],
    ]


@pytest.mark.integration
def test_multiple_files(runner, test_root_path, snapshot):
    query_file = f"{test_root_path}/test_data/sql_multi_queries.sql"
    result = runner.invoke_with_connection(
        [
            "sql",
            "-f",
            query_file,
            "-f",
            f"{test_root_path}/test_data/empty.sql",
            "-f",
            query_file,
        ]
    )

    assert result.exit_code == 0
    assert result.output == snapshot


@pytest.mark.integration
def test_multi_queries_where_one_of_them_is_failing(runner, test_root_path):
    result = runner.invoke_with_connection_json(
        ["sql", "-q", f"select 1; select 2; select foo; select 4", "--format", "json"],
    )
    assert result.exit_code == 1

    assert '"1"  :   1' in result.output
    assert '"2"  :   2' in result.output
    assert "invalid identifier 'FOO'" in result.output


@pytest.mark.integration
def test_multi_input_from_stdin(runner, test_root_path):
    result = runner.invoke_with_connection_json(
        [
            "sql",
            "-i",
        ],
        input="select 1, 2, 3 union select 4, 5, 6; select 42",
    )
    assert result.exit_code == 0
    assert result.json == [
        [{"1": 1, "2": 2, "3": 3}, {"1": 4, "2": 5, "3": 6}],
        [{"42": 42}],
    ]


def _round_values(results):
    for result in results:
        for k, v in result.items():
            result[k] = round(v, 2)
    return results


def _round_values_for_multi_queries(results):
    return [_round_values(r) for r in results]


@pytest.mark.integration
@mock.patch("snowflake.cli._app.printing._get_table")
@mock.patch("snowflake.cli._app.printing.Live")
def test_queries_are_streamed_to_output(
    _, mock_get_table, runner, capsys, test_root_path
):
    log = []

    def _log(
        *args,
    ):
        log.append((time.monotonic(), *args))

    # We mock add_row by a function that logs call time, later we compare call times
    # to make sure those were separated in time
    mock_get_table().add_row = _log

    runner.invoke_with_connection(
        [
            "sql",
            "-q",
            "select 13; select system$wait(10);",
        ],
    )

    assert len(log) == 2
    (time_0, query_0), (time_1, query_1) = log

    assert query_0 == "13"
    assert time_1 - time_0 >= 10.0
    assert "waited 10 seconds" in query_1


@pytest.mark.integration
@pytest.mark.parametrize(
    "query, expected",
    (
        pytest.param(
            "select 1; -- trailing comment\n",
            [
                {"1": 1},
            ],
            id="single query",
        ),
        pytest.param(
            "select 1; --comment\n select 2; \n -- trailing comment\n",
            [
                [
                    {"1": 1},
                ],
                [
                    {"2": 2},
                ],
            ],
        ),
    ),
)
def test_trailing_comments_queries(runner, query, expected, test_root_path):
    result = runner.invoke_with_connection_json(
        ["sql", "-q", query, "--format", "JSON"]
    )
    assert result.exit_code == 0
    assert result.json == expected, result.json


@pytest.mark.integration
def test_sql_execute_query_prints_query(runner):
    result = runner.invoke_with_connection(
        ["sql", "-q", "select 1 as A; select 2 as B"]
    )

    assert result.exit_code == 0, result.output
    assert "select 1 as A" in result.output
    assert "select 2 as B" in result.output


@pytest.mark.integration_experimental
def test_sql_large_lobs_in_memory_tables(runner):
    table_name = ObjectNameProvider(
        "table_with_default"
    ).create_and_get_next_object_name()
    result = runner.invoke_with_connection(
        [
            "sql",
            "-q",
            f"create or replace table {table_name}(x int, v text default x::varchar);"
            f"select get_ddl('table', '{table_name}');"
            f"drop table {table_name};",
        ]
    )

    assert "VARCHAR(134217728)" in result.output


@pytest.mark.integration
@pytest.mark.parametrize(
    "template",
    [
        "<% ctx.env.monty_python %>",
        "&{ ctx.env.monty_python }",
    ],
)
def test_sql_with_variables_from_project(runner, project_directory, template):
    with project_directory("sql"):
        result = runner.invoke_with_connection_json(
            [
                "sql",
                "-q",
                f"select '{template}' as var",
            ]
        )
        assert result.exit_code == 0, result.output
        assert result.json == [{"VAR": "Knights of Nii"}]
