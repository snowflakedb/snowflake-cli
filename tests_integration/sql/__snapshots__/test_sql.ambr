# serializer version: 1
# name: test_multiple_files
  '''
  select round(ln(1), 4);
  +-----------------+
  | ROUND(LN(1), 4) |
  |-----------------|
  | 0.0             |
  +-----------------+
  select round(ln(10), 4);
  +------------------+
  | ROUND(LN(10), 4) |
  |------------------|
  | 2.3026           |
  +------------------+
  select round(ln(100), 4);
  +-------------------+
  | ROUND(LN(100), 4) |
  |-------------------|
  | 4.6052            |
  +-------------------+
  select round(ln(1), 4);
  +-----------------+
  | ROUND(LN(1), 4) |
  |-----------------|
  | 0.0             |
  +-----------------+
  select round(ln(10), 4);
  +------------------+
  | ROUND(LN(10), 4) |
  |------------------|
  | 2.3026           |
  +------------------+
  select round(ln(100), 4);
  +-------------------+
  | ROUND(LN(100), 4) |
  |-------------------|
  | 4.6052            |
  +-------------------+
  
  '''
# ---
# name: test_sql_source_command_from_user_input
  '''
  select 1;
  +---+
  | 1 |
  |---|
  | 1 |
  +---+
  select 42;
  +----+
  | 42 |
  |----|
  | 42 |
  +----+
  select 3;
  +---+
  | 3 |
  |---|
  | 3 |
  +---+
  
  '''
# ---
